File: ports/adf2mp3.rs

// ================================================================================================
// -*- Rust -*-
// File: adf2mp3.rs
// Author: Guilherme R. Lampert
// Created on: 08/03/15
// Brief: GTA Vice City ADF to MP3 converter (Rust version of the original C++ code).
//
// This code is public domain. I claim no copyrights over it.
// No warranty is offered or implied; use it at your own risk.
//
// ================================================================================================

// Currently (rustc 1.0.0-nightly), these modules are all unstable
// and would generate warnings without these suppressions.
#![feature(io)]
#![feature(path)]
#![feature(collections)]

use std::fs::File;
use std::fs::Metadata;
use std::path::Path;
use std::io::prelude::*;

// ========================================================

// The GTA Vice City ADF files are MP3 files that
// had each byte XORed with this magic constant.
// 34 is 22 in hexadecimal and 42 in octal...
//
// Not sure who figured this out, but I got this
// info from the Xentax File Format Wiki.
//
const GTA_MAGIC: u8 = 34;

// Process the input file in chunks of this size in bytes.
// The chunk buffer is allocated dynamically (it is a Vec)
// so this can be a fairly large value.
//
const CHUNK_SIZE: usize = 8192;

// ========================================================

fn print_help_text(program_name: &str) {
	println!("");
	println!("Usage:");
	println!("$ {} <input_file> [output_file]", program_name);
	println!("  Runs the tool normally. If the output filename is not provided");
	println!("  the input filename is used but the extension is replaced with '.mp3'.");
	println!("");
	println!("Usage:");
	println!("$ {} --help | -h", program_name);
	println!("  Prints this help text.");
	println!("");
}

fn remove_extension(filename: &str) -> String {
	String::from_str(Path::new(filename).file_stem()
		.unwrap().to_str().unwrap())
}

fn open_file(filename: &str) -> (File, Metadata) {
	let path = Path::new(filename);

	// `open()` is read-only by default.
	let file_handle = match File::open(&path) {
		Err(why) => panic!("Couldn't open file {}: {}", path.display(), why.description()),
		Ok(file_handle) => file_handle,
	};

	let file_info = match file_handle.metadata() {
		Err(why) => panic!("Unable to get metadata for {}: {}", path.display(), why.description()),
		Ok(file_info) => file_info,
	};

	// Return a pair of file handle and its metadata:
	(file_handle, file_info)
}

fn create_file(filename: &str) -> File {
	let path = Path::new(filename);

	// `create()` is write-only by default.
	match File::create(&path) {
		Err(why) => panic!("Couldn't create file {}: {}", path.display(), why.description()),
		Ok(file_handle) => file_handle,
	}
}

fn process_files(input_filename: &str, output_filename: &str) {
	// Source file (ADF):
	let in_file_tuple = open_file(input_filename);
	let in_file_len   = in_file_tuple.1.len() as usize; // Index 1: Metadata
	let mut in_file   = in_file_tuple.0;                // Index 0: File

	// Destination file (MP3):
	let mut out_file = create_file(output_filename);

	// Chunk processing loop:
	let mut bytes_processed: usize = 0;
	while bytes_processed != in_file_len {
		let bytes_left = in_file_len - bytes_processed;
		let bytes_this_iteration = std::cmp::min(CHUNK_SIZE, bytes_left);

		let mut chunk = vec![0u8; bytes_this_iteration];

		match in_file.read(chunk.as_mut_slice()) {
			Err(why) => panic!("Failed to read input file! {}", why.description()),
			Ok(_) => {},
		};

		for byte in &mut chunk {
			*byte ^= GTA_MAGIC;
		}

		match out_file.write(chunk.as_mut_slice()) {
			Err(why) => panic!("Failed to write output file! {}", why.description()),
			Ok(_) => {},
		};

		bytes_processed += bytes_this_iteration;
	}
}

fn main() {
	// Fetch command line:
	let cmd_args: Vec<_> =
		std::env::args()
		.map(|x| x.to_string())
		.collect();

	// Too few command line args, exit:
	if cmd_args.len() < 2 {
		println!("Not enough arguments!");
		print_help_text(&cmd_args[0]);
		return;
	}

	// Print help and exit if "-h" or "--help" present in cmd line:
	if cmd_args.len() > 1 && (cmd_args[1] == "-h" || cmd_args[1] == "--help") {
		print_help_text(&cmd_args[0]);
		return;
	}

	//
	// Normal execution path:
	//
	let input_filename;
	let mut output_filename;

	if cmd_args.len() >= 3 {
		// input_filename + output_filename:
		input_filename  = cmd_args[1].clone();
		output_filename = cmd_args[2].clone();
	} else {
		// Just input_filename:
		input_filename  = cmd_args[1].clone();
		output_filename = String::new();
	}

	// Replace ".adf" extension of source filename with ".mp3" and use
	// it for the output if no explicit filename was provided.
	if output_filename.is_empty() {
		output_filename = remove_extension(&input_filename) + ".mp3";
	}

	process_files(&input_filename, &output_filename);
}

File: adf2mp3.cpp
// ================================================================================================
// -*- C++ -*-
// File: adf2mp3.cpp
// Author: Guilherme R. Lampert
// Created on: 23/12/14
// Brief: GTA Vice City ADF to MP3 converter.
//
// This code is public domain. I claim no copyrights over it.
// No warranty is offered or implied; use it at your own risk.
//
// ================================================================================================

// Standard C++ includes:
#include <algorithm>
#include <fstream>
#include <iostream>
#include <stdexcept>
#include <string>
#include <vector>

// Standard C / POSIX includes:
#include <cerrno>
#include <cstdlib>
#include <sys/types.h>
#include <sys/stat.h>

namespace adf2mp3
{
	using ubyte = unsigned char;
	using ulong = unsigned long;

	//
	// The GTA Vice City ADF files are MP3 files that
	// had each byte XORed with this magic constant.
	// 34 is 22 in hexadecimal and 42 in octal...
	//
	// Not sure who figured this out, but I got this
	// info from the Xentax File Format Wiki.
	//
	const ubyte GtaMagic = 34;

	//
	// Process the input file in chunks of this size in bytes.
	// The chunk buffer is allocated dynamically (it is a std::vector)
	// so this can be a fairly large value.
	//
	const ulong ChunkSize = 8192;

	// ========================================================

	struct FileStats
	{
		ulong fileLength;   // File length in bytes.
		bool  isDirectory;  // True if the path provided pointed to a directory.
		bool  isNormalFile; // True is the path provided pointed to a file.
	};

	FileStats fileStatsForPath(const std::string & pathname)
	{
		errno = 0; // Clear the errno global, just in case.
		struct stat statBuf;

		if (stat(pathname.c_str(), &statBuf) != 0)
		{
			throw std::runtime_error("Path '" + pathname + "': " + std::string(std::strerror(errno)));
		}

		FileStats fs;
		fs.fileLength = static_cast<ulong>(statBuf.st_size);

#if _WIN32
		fs.isDirectory = (statBuf.st_mode & _S_IFDIR) ? true : false;
		fs.isNormalFile = (statBuf.st_mode & _S_IFREG) ? true : false;
#else // !_WIN32
		fs.isDirectory = S_ISDIR(statBuf.st_mode) ? true : false;
		fs.isNormalFile = S_ISREG(statBuf.st_mode) ? true : false;
#endif // _WIN32

		return fs;
	}

	std::string removeExtension(const std::string & filename)
	{
		const auto lastDot = filename.find_last_of('.');
		if (lastDot == std::string::npos)
		{
			return filename;
		}
		return filename.substr(0, lastDot);
	}

	void printHelpText(const char * programName)
	{
		std::cout << "\n";
		std::cout << "Usage:\n";
		std::cout << "$ " << programName << " <input_file> [output_file]\n";
		std::cout << "  Runs the tool normally. If the output filename is not provided\n" <<
			"  the input filename is used but the extension is replaced with '.mp3'.\n";

		std::cout << "\n";
		std::cout << "Usage:\n";
		std::cout << "$ " << programName << " --help | -h\n";
		std::cout << "  Prints this help text.\n";
		std::cout << "\n";
	}

	void handleCommandLine(int argc, const char * argv[],
		std::string & inFileName, std::string & outFileName,
		bool & printHelpOnly, bool & missingArgs)
	{
		if (argc < 2)
		{
			missingArgs = true;
			return;
		}

		if (std::strcmp(argv[1], "--help") == 0 || std::strcmp(argv[1], "-h") == 0)
		{
			printHelpOnly = true;
			return;
		}

		// input_file + output_file
		if (argc >= 3)
		{
			inFileName = argv[1];
			outFileName = argv[2];
		}
		else // Just input_file
		{
			inFileName = argv[1];
			outFileName.clear();
		}

		printHelpOnly = false;
		missingArgs = false;
	}

	void processChunk(std::vector<ubyte> & chunk, ulong numBytes)
	{
		for (ulong b = 0; b < numBytes; ++b)
		{
			chunk[b] ^= GtaMagic;
		}
	}

	void processFiles(const std::string & inFileName, const std::string & outFileName)
	{
		const FileStats fileStats = fileStatsForPath(inFileName);

		if (fileStats.fileLength == 0)
		{
			throw std::runtime_error("Input file is empty!");
		}

		if (fileStats.isDirectory)
		{
			throw std::runtime_error("Input file '" + inFileName + "' is a directory!");
		}

		std::ifstream inFile(inFileName, std::ifstream::in | std::ifstream::binary);
		std::ofstream outFile(outFileName, std::ofstream::out | std::ofstream::binary);

		ulong bytesProcessed = 0;
		std::vector<ubyte> chunk(ChunkSize, 0);

		while (bytesProcessed != fileStats.fileLength)
		{
			const ulong bytesLeft = fileStats.fileLength - bytesProcessed;
			const ulong bytesThisIteration = std::min<ulong>(chunk.size(), bytesLeft);

			inFile.read(reinterpret_cast<char *>(chunk.data()), bytesThisIteration);
			processChunk(chunk, bytesThisIteration);
			outFile.write(reinterpret_cast<char *>(chunk.data()), bytesThisIteration);

			bytesProcessed += bytesThisIteration;
		}
	}

	void run(int argc, const char * argv[])
	{
		bool printHelpOnly, missingArgs;
		std::string inFileName, outFileName;
		handleCommandLine(argc, argv, inFileName, outFileName, printHelpOnly, missingArgs);

		if (printHelpOnly)
		{
			printHelpText(argv[0]);
			return;
		}

		if (missingArgs)
		{
			std::cout << "Not enough arguments!\n";
			printHelpText(argv[0]);
			return;
		}

		// Replace .adf extension of source file with .mp3 and use
		// it for the output if no explicit filename was provided.
		if (outFileName.empty())
		{
			outFileName = removeExtension(inFileName) + ".mp3";
		}

		processFiles(inFileName, outFileName);
	}
} // namespace adf2mp3 {}

int main(int argc, const char * argv[])
{
	try
	{
		adf2mp3::run(argc, argv);
		return 0;
	}
	catch (std::exception & e)
	{
		std::cerr << "ERROR: " << e.what() << std::endl;
		return EXIT_FAILURE;
	}
}